### DISCUSSION TITLE
How to read parameters from an html form POST request?

### DISCUSSION THREAD

**Discussion by mcordova1967** [uNetworking/uWebSockets#1467] [Q&A]
I understand that this server can be used as a web server too, but I cannot find the APIs to read post parameters sent as application/x-www-form-urlencoded - is it supported? If not, what's the API to get the payload of a POST request and proceed to parse it myself? thank you! Martin.

**Comment by whitwhoa** [comment]
I've been looking for the same thing and what I found was the `onData` callback. You would need to parse the payload yourself from the string buffer. If what you're passing can be json encoded, then this should be trivial (what I'm working on now). Of course, if someone else knows of a better way of doing this, by all means...please share! :) Below is a simple example of reading POST data:

```
uWS::App().post("/api/signup", [&](auto *res, auto */*req*/) {

	//std::cout << "Request accepted" << std::endl;

	/* We have to attach an abort handler for us to be aware
	* of disconnections while we perform async tasks */
	res->onAborted([=]() {
		std::cout << "HTTP socket was closed before we request completed" << std::endl;
	});

	res->onData([res](std::string_view data, bool last) {

		std::cout << "Got data: " << data.length() << ", it's the last: " << last << std::endl;

		if (last)
		{
                        // do something with post data here

			res->end(data); // just echo the submitted post data back to client
		}
		
	});
	

}).listen(3000, [](auto *listen_socket) {

	if (listen_socket) 
	{
		std::cout << "Listening on port " << 3000 << std::endl;
	}

}).run();

std::cout << "Failed to listen on port 3000" << std::endl;
```

**Reply by ghost** [reply]
Isn't this what the MultiPart parser is for? https://github.com/uNetworking/uWebSockets/blob/master/src/Multipart.h

**Reply by mcordova1967** [reply]
Thank you @whitwhoa, that sheds some light

**Reply by whitwhoa** [reply]
@alexhultman I'm confused (obviously). After looking through the Multipart.h file it appears there's a block that returns if the content type does not contain a substring of `multipart/`. Therefore in the case of submitting data with content types of `application/x-www-form-urlencoded` or `application/json` would parameters not be ignored? Sorry if I'm not understanding correctly what you had meant. BUT I most certainly would like to insure I'm doing things the correct way. If the above is incorrect are there examples or documentation you could point me to so that I might better understand the correct way of obtaining POST (or GET parameters even) form parameters?
